planutils/package_installation.py

```diff
@@ -0,0 +1,82 @@
+
+import json, os, glob, subprocess
+
+from planutils import settings
+
+PACKAGES = {}
+
+CUR_DIR = os.path.dirname(os.path.abspath(__file__))
+
+for conf_file in glob.glob(os.path.join(CUR_DIR, 'packages', '*')):
+    base = os.path.basename(conf_file)
+    if base not in ['README.md', 'TEMPLATE']:
+        with open(os.path.join(conf_file, 'manifest.json'), 'r') as f:
+            config = json.load(f)
+        assert base not in PACKAGES, "Error: Duplicate package config -- %s" % base
+        PACKAGES[base] = config
+
+
+def check_installed(target):
+    return target in settings.load()['installed']
+
+
+def uninstall(target):
+    if 'list' == target:
+        print("\nInstalled Packages:")
+        for p in PACKAGES:
+            if check_installed(p):
+                print(" - %s: %s" % (p, PACKAGES[p]['name']))
+        print()
+        return
+    
+    if target not in PACKAGES:
+        print("Error: Package not found -- %s" % target)
+        return
+
+    if not check_installed(target):
+        print("%s isn't installed." % target)
+    else:
+        print ("Uninstalling %s..." % target)
+        s = settings.load()
+        # map a package to all those that depend on it
+        deps = {}
+        for p in s['installed']:
+            for dep in PACKAGES[p]['dependencies']:
+                if dep not in deps:
+                    deps[dep] = set()
+                deps[dep].add(p)
+
+        if target in deps and len(deps[target]) > 0:
+            print("Error: Package is required for the following: %s" % ', '.join(deps[target]))
+            return
+        subprocess.call('./uninstall', cwd=os.path.join(CUR_DIR, 'packages', target))
+        s['installed'].remove(target)
+        settings.save(s)
+
+
+
+def install(target):
+    if 'list' == target:
+        print("\nPackages:")
+        for p in PACKAGES:
+            print(" - %s: %s (installed: %s)" % \
+                    (p, PACKAGES[p]['name'], str(check_installed(p))))
+        print()
+        return
+
+    if target not in PACKAGES:
+        print("Error: Package not found -- %s" % target)
+        return
+
+    for dep in PACKAGES[target]['dependencies']:
+        if not check_installed(dep):
+            install(dep)
```

Comment:
At this point, some output to the user would be nice. Something like: "x depends on y1, y2, and y3 [install all/cancel]"

Reply:
Maybe this and the `uninstall` command should take lists of packages and print a list of changes about to happen with the option to cancel (like `apt` does).